### PR TITLE
MSI-1190: Add ability to create shipment with product with custom options.

### PR DESCRIPTION
**Pull request by nmalevanec**
### Description

If product has custom options with it's own sku defined, shipment(invoice, order) item has sku = $productSku + $optionSku. 
So in:
https://github.com/magento-engcom/msi/blob/1ad4abd737236683ad1f9991928d04278dd14876/app/code/Magento/InventoryShipping/Model/GetItemsToDeduct.php#L64-L66

https://github.com/magento-engcom/msi/blob/1ad4abd737236683ad1f9991928d04278dd14876/app/code/Magento/InventoryShipping/Model/InventoryRequestFromInvoiceFactory.php#L98-L100

and

https://github.com/magento-engcom/msi/blob/1ad4abd737236683ad1f9991928d04278dd14876/app/code/Magento/InventoryShipping/Model/InventoryRequestFromOrderFactory.php#L72-L74

items for deduction will have incorrect sku, which leads to inability to retrieve stock Item configuration for given product.

---

Pull request with additional "create shipment" step for CreateOrderBackendPartOneTest:
magento/magento2#16153

### Fixed Issues (if relevant)

1. magento-engcom/msi#1190: Fix Functional Test Magento\Sales\Test\TestCase\CreateOrderBackendPartOneTest::test with data set "CreateOrderBackendTestVariation1_0"

### Manual testing scenarios

1. Login to admin area.
2. Create virtual and simple products with custom options, which have their own sku
image
3. Navigate to storefront and create orders with created products.
4. Navigate back to admin area.
5. Invoice order with virtual product.
6. Check, invoice created successfully.
7. Navigate to Catalog > Products.
8. Check, virtual product quantity has decreased correctly.
9. Create shipment for order with simple product.
10. Check, shipment created successfully.
11. Navigate to Catalog > Products.
12. Check, simple product quantity has decreased correctly.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/Product/AddProductSkuToBuyRequest.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+     * @param mixed $value
+     * @param int|Product $product
+     * @return array
+     * @throws \Magento\Framework\Exception\NoSuchEntityException in case requested product doesn't exist.
+     */
+    public function beforeAddCustomOption(Product $subject, $code, $value, $product = null) :array
+    {
+        if ($code === 'info_buyRequest') {
+            $sku = $this->getSkusByProductIds->execute([$subject->getId()])[$subject->getId()];
+            $value = $this->serializer->unserialize($value);
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/Product/AddProductSkuToBuyRequest.php]: $value could be mixed value, need to add one more check

File: dev/tests/integration/testsuite/Magento/Downloadable/Model/Product/TypeTest.php [hunk 2]
```diff
@@ -250,8 +250,8 @@ public function testCheckProductBuyState()
         $linksFactory = $this->objectManager
             ->get(\Magento\Downloadable\Model\ResourceModel\Link\CollectionFactory::class);
         $allLinksIds = $linksFactory->create()->addProductToFilter($product->getEntityId())->getAllIds();
-        $this->assertEquals(
-            '{"qty":23,"links":["' . implode('","', $allLinksIds). '"]}',
+        $this->assertRegExp(
```
Comment (naydav) [inline dev/tests/integration/testsuite/Magento/Downloadable/Model/Product/TypeTest.php]: We don't need to rely on concrete implementation, use `\Magento\Framework\Serialize\Serializer\Json`

File: app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/Product/AddProductSkuToBuyRequest.php (hunk 1 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/Product/AddProductSkuToBuyRequest.php]: Done

File: dev/tests/integration/testsuite/Magento/Downloadable/Model/Product/TypeTest.php (hunk 2 above)
Comment (nmalevanec) [reply dev/tests/integration/testsuite/Magento/Downloadable/Model/Product/TypeTest.php]: Reworked

**Comment by ishakhsuvarov** [comment]
Closing this PR for now, #2012 seems to solve this problem already

**Pull request closed without merging**